### PR TITLE
Tock 2.0: implement Upcall swapping restrictions (v4)

### PR DESCRIPTION
**Pull request by bradjc**
### Pull Request Overview

This is an alternative approach to #2462 that ensures apps always get back the previous upcall on a subscribe call.

This implementation works by handling all upcall storage and swapping entirely in the core kernel. The core kernel creates its own data structure as a process grant and stores upcalls there. Since the core kernel is entirely responsible, we can ensure that upcall swapping is always handled correctly.

Capsules now no longer store upcalls. Instead they use an API like:

```rust
self.grants.enter(processid, |process_grant, upcalls| {
    ... // use the grant

    upcalls.schedule_upcall(upcall_id, arg0, arg1, arg2);
}
```

Other ramifications of this change:

- `Driver` now includes a `allocate_grant()` function that capsules have to implement. This provides the core kernel a method for forcing a grant allocation so that it can store an upcall if the process calls subscribe before the grant has been allocated.
- Grants need to be identified with the driver number of the capsule that is going to use them. This allows the kernel to match the correct upcall on a subscribe call.
- Grant<T> is now `Grant<T, NUM_UPCALLS>` so capsules can tell the core kernel how many upcalls they need stored.

### Testing Strategy

Tested apps on hail and imix.

### TODO or Help Wanted

- [x] Move pointer manipulation code out of process_standard.rs
- [x] Unsafe documentation
- [x] Cleanup
- [x] Fix IPC to make `NUM_CALLBACKS` tied to `NUM_PROCESSES` again, and to address that IPC did not follow TRD 104 (additional actions were taken in response to a subscribe)

### Documentation Updated

- [x] Updated the relevant files in `/docs`, or no updates are required.

### Formatting

- [x] Ran `make prepush`.

### CODE REVIEW

**Comment by alexandruradovici** [comment]
Nice implementation, from the capsule writer point of view this seems to make driver writing easier.

I have two concerns though.

What happens if a capsule has two or more grants? My assumption based on the `driver_num_to_grant_num` implementation is that the first one created (or the first that is actually allocated if the capsule only allocates some of them) is used for upcalls.

As the driver number is now used in two places, the `with_driver` function and the `create_grant`, it is easy to make a mistake and supply two different driver numbers, which would result in upcalls silently failing and registering upcalls for another capsule.

**Comment by hudson-ayers** [comment]
Currently, this PR reduce code size on Imix by ~300 bytes.

**Comment by hudson-ayers** [comment]
> I have two concerns though.
> 
> What happens if a capsule has two or more grants? My assumption based on the `driver_num_to_grant_num` implementation is that the first one created (or the first that is actually allocated if the capsule only allocates some of them) is used for upcalls.

Good catch -- we need to bring back the check from #2462 that there is only one grant per driver. That is done now. I think we might be able to get away with just doing a check when each grant is allocated, rather than having to use a separate table, but using the existing solution for now.

> As the driver number is now used in two places, the `with_driver` function and the `create_grant`, it is easy to make a mistake and supply two different driver numbers, which would result in upcalls silently failing and registering upcalls for another capsule.

Yeah, this problem also exists for #2462 . But I just added some checks to this PR so that this mistake will not lead to registering upcalls for another capsule, and instead lead to an error. We could replace this with a panic for it not to be silent

**Comment by alexandruradovici** [comment]
> > I have two concerns though.
> > What happens if a capsule has two or more grants? My assumption based on the `driver_num_to_grant_num` implementation is that the first one created (or the first that is actually allocated if the capsule only allocates some of them) is used for upcalls.
> 
> Good catch -- ~we need to bring back the check from #2462 that there is only one grant per driver. That is done now. I think we might be able to get away with just doing a check when each grant is allocated, rather than having to use a separate table, but using the existing solution for now.~
> I changed this to check at the time of grant allocation if there are multiple grants with the same driver_num.
> 
> > As the driver number is now used in two places, the `with_driver` function and the `create_grant`, it is easy to make a mistake and supply two different driver numbers, which would result in upcalls silently failing and registering upcalls for another capsule.
> 
> Yeah, this problem also exists for #2462 . But I just added some checks to this PR so that this mistake will not lead to registering upcalls for another capsule, and instead lead to an error. We could replace this with a panic for it not to be silent.

You are right, but it still bugs me that the driver number has to be supplied in two places. Definitely the check helps.

**Comment by bradjc** [comment]
> You are right, but it still bugs me that the driver number has to be supplied in two places. Definitely the check helps.

Unfortunately I think having a one-to-one mapping of grant to driver_num is fundamental, and the `Grant` type needs to know the driver_num. If we changed the design to remove the `with_driver()` mapping function we could hard-code driver numbers for capsules, but then we restrict how board authors can customize their kernel.

File: kernel/src/upcall.rs
Comment (hudson-ayers) [reply kernel/src/upcall.rs:18]: In #2462 these were ported to u32

File: kernel/src/grant.rs
Comment (bradjc) [inline kernel/src/grant.rs:1009]: Are these changes supposed to be here?

**Comment by bradjc** [comment]
The hail app now works on hail!

File: kernel/src/grant.rs
Comment (hudson-ayers) [reply kernel/src/grant.rs:1009]: nope, thats a find/replace mistake, sorry

File: kernel/src/sched.rs
Comment (hudson-ayers) [inline kernel/src/sched.rs:959]: we should probably check if the grant is allocated before calling this

Comment (bradjc) [reply kernel/src/sched.rs:959]: We do on line 915.

File: kernel/src/sched.rs
Comment (bradjc) [inline kernel/src/sched.rs:905]: @phil-levis Can you read through the subscribe match block, and check that the logic flow and comments match what TRD104 promises?

File: kernel/src/sched.rs
Comment (hudson-ayers) [reply kernel/src/sched.rs:959]: Do we? This is what I see for line 915:
```
                // First check if `upcall_ptr` is null. A null upcall pointer
                // represents the special "unsubscribe" operation.
                let ptr = NonNull::new(upcall_ptr);
```

Comment (bradjc) [reply kernel/src/sched.rs:959]: Ha that's what I get for doing it on my phone. That should be "951". Basically we first try the subscribe, and only if it fails do we try to do the grant allocation and eventually retry the subscribe.

File: kernel/src/upcall.rs
Comment (alevy) [reply kernel/src/upcall.rs:18]: yeah... `usize` if they are register sized and `u32` if they are 4-byte words (which happen to be the same on all of our platforms). Since an upcall is something that models the interaction over a system call boundary, it seems to me it should be a `usize`. E.g., "how many subscribe numbers can there be?" answer: "the max value that can fit in a CPU register", not, "2^32".

**Comment by bradjc** [comment]
Ok I believe all in code TODOs have been resolved, and all new `unsafe` usages are documented. I updated the new boards changes to use the new create_grant signature.

The remaining issue is that IPC is broken because it can no longer do any operations as a result of a subscribe.

**Comment by hudson-ayers** [comment]
I pushed the IPC fix. The main breakage was that the number of upcalls now needs to be passed as a const, but for IPC the number of upcalls is `NUM_PROCS + 1`, and `min_const_generics` does not let you do arithmetic on passed consts. So for now we need another generic parameter, unfortunately.

I investigated everything that the existing subscribe implementation did, and it was actually just buffer swapping with some extra checks (which allowed the driver to reject subscribes). As far as I can tell the only check that has been removed by letting the kernel handle all this is a call to `lookup_app_by_identifier()`, called when an app subscribes to a service by registering an upcall in the slot associated with that service. This check was used to verify a process actually existed in that slot before trying to register an upcall for that service. Now, if an app tries to register an upcall for a service corresponding to a non-existent process, it will not get an error. I don't think this is a huge deal, personally.

**Comment by bradjc** [comment]
Ok, the IPC issues have been resolved, and I updated the capsule porting guide. I think this is now a viable PR to implement the upcall swapping restrictions.

**Comment by hudson-ayers** [comment]
I have tested this PR with blink and with the IPC apps on Imix

**Comment by hudson-ayers** [comment]
Also, the final version of this PR reduces code size on Imix by 1500 bytes!

File: kernel/src/sched.rs
Comment (phil-levis) [reply kernel/src/sched.rs:923]: Will do.

**Comment by bradjc** [comment]
One issue that comes up with this approach, or any approach that tags a Grant with a driver_num, is that it is possible to have a mismatch between the syscall driver_num and the grant_driver num, and depending on what userspace does this could manifest in some very strange behavior for userspace processes that would not be immediately obvious to developers.

One thought is to add a `panic()` if the kernel detects a mismatch. This would be palatable from a tock perspective because this would be the fault of a faulty main.rs file, which we consider trusted code anyway. However, I do not think this is possible, because the mismatch case would present identically to the case where a capsule maliciously implements `Driver.allocate_grant()` (i.e. does not allocate its grant when asked). In both cases, the kernel has to handle a subscribe call where it does not find a matching grant to store it in. But if we panic, then a capsule can cause a kernel panic entirely by how it implements one function (allocate_grant()), in a way that a capsule is technically allowed to do.

Another thought is to make the mistake harder to make by embedding the default driver_num in components themselves. Since `with_driver()` is in our experience always implemented correctly and using the default DRIVER_NUMs, this would help ensure that the same driver_num is used in the syscall mapping and in the create_grant() call. However, by hiding the fact that the driver_num is stored in the capsule, this could cause other problems when developers aren't aware the driver_num is in the capsule and they must ensure it matches. Also, the version where driver_num is passed to components has been around for a while in PRs. Ultimately, we can change our mind on how to minimize the chance of mixing up the driver_num, but it seems like it is worth seeing if this is a common bug before trying too hard to solve it.

File: kernel/src/syscall.rs
Comment (phil-levis) [inline kernel/src/syscall.rs:280]: Why is this `mut` rather than const?

File: kernel/src/syscall.rs
Comment (phil-levis) [inline kernel/src/syscall.rs:283]: Why is this `mut` rather than const?

**Comment by lschuermann** [comment]
I agree we want to get this in quickly, but should this be `last-call` already? It's a major change to the core kernel, I would've liked to have an in depth look over the weekend.

File: kernel/src/process_standard.rs
Comment (phil-levis) [inline kernel/src/process_standard.rs:1412]: `*const`

**Comment by jrvanwhy** [comment]
> I agree we want to get this in quickly, but should this be `last-call` already? It's a major change to the core kernel, I would've liked to have an in depth look over the weekend.

We were planning to wait until at least Monday to merge.

File: kernel/src/grant.rs
Comment (phil-levis) [inline kernel/src/grant.rs:169]: `UpcallTable`

File: doc/Porting_v1_Capsules_to_v2.md
Comment (phil-levis) [inline doc/Porting_v1_Capsules_to_v2.md:278]: Explain why, how V2 subscribe is different.

**Review by phil-levis — APPROVED** [review]

File: kernel/src/grant.rs
Comment (phil-levis) [inline kernel/src/grant.rs:713]: From call:

GrantMemory -> GrantData
UpcallMemory -> GrantUpcallTable

File: kernel/src/process_standard.rs
Comment (bradjc) [reply kernel/src/process_standard.rs:1412]: This is a pointer to mutable memory, as it points to where we store the upcalls and grant object.

File: kernel/src/grant.rs
Comment (bradjc) [reply kernel/src/grant.rs]: Changed UpcallMemory, other change is in 2651.

File: doc/Porting_v1_Capsules_to_v2.md
Comment (bradjc) [reply doc/Porting_v1_Capsules_to_v2.md]: Updated.

File: kernel/src/process_standard.rs
Comment (phil-levis) [reply kernel/src/process_standard.rs:1412]: Ah, right. Sorry, my pattern-match wasn't selective enough.

File: kernel/src/grant.rs
Comment (lschuermann) [inline kernel/src/grant.rs:879]: Can this be a `slice::from_raw_parts_mut` here?

File: kernel/src/grant.rs
Comment (lschuermann) [inline kernel/src/grant.rs:667]: Why do we pass a mutable reference to `GrantUpcallTable`? My fear is that users are going to be able to `mem::swap` this potentially, though this probably is (a) not possible because of non-reentrant Grants and (b) not a concern given the passed `GrantUpcallTable` reference is dropped afterwards. Still, if there's no good reason for it to be mutable, it shouldn't be.

File: kernel/src/grant.rs
Comment (lschuermann) [inline kernel/src/grant.rs:981]: Here as well.

**Review by lschuermann — DISMISSED** [review]
These changes look good and seem to improve semantics w.r.t. driver implementations. I had a very close look at the kernel code and could not find any unsafety / unsoundness. I have not tested this at runtime, nor its performance impact.

File: kernel/src/grant.rs
Comment (ppannuto) [inline kernel/src/grant.rs:217]: Should this return a `Result<(), Error>` rather than a bool?

It's a bit annoying from a verbosity perspective, as it results in a boilerplate `upcalls.schedule_upcall(..)` -> `upcalls.schedule_upcall(..).ok()` for most callers to silence the ignored `Result`, but it does seem clearer than the `bool`, and there is some benefit to asking capsule authors to consider briefly whether they care about the process not actually being notified.

File: kernel/src/grant.rs
Comment (ppannuto) [inline kernel/src/grant.rs:550]:
```suggestion
                // 2. The pointer is correctly aligned because we calculated the
```

File: kernel/src/grant.rs
Comment (ppannuto) [inline kernel/src/grant.rs:572]:
```suggestion
                        // least aligned to `usize` bytes.
```
?

**Review by ppannuto — COMMENTED** [review]
Read over `kernel/` and `doc/` pretty carefully. I think this all looks good; a few small doc cleanups and one interface question.

File: kernel/src/grant.rs
Comment (alexandruradovici) [inline kernel/src/grant.rs:94]:
```suggestion
//!                         │   driver_num: usize          │
//!                         │   grant_num: usize          │
```

File: kernel/src/sched.rs
Comment (alexandruradovici) [inline kernel/src/sched.rs:989]: It might be a good idea to show a debug message if the driver wrongly allocates a wrong grant. https://github.com/tock/tock/pull/2654

**Review by alexandruradovici — COMMENTED** [review]
I have a small documentation update and a debug suggestion.

File: kernel/src/grant.rs
Comment (hudson-ayers) [reply kernel/src/grant.rs:879]: I'm not sure what you mean -- it already is a `slice::from_raw_parts_mut` here? Is the concern that using it this way is unsafe?

File: kernel/src/grant.rs
Comment (hudson-ayers) [reply kernel/src/grant.rs]: I agree this is not needed, and pushed a commit changing it to a non-mutable reference

File: kernel/src/grant.rs
Comment (hudson-ayers) [reply kernel/src/grant.rs:217]: I just ported the existing interface for `schedule()`. I think you're right that a `Result` would be better, but I might prefer for that to be submitted as a follow up PR, since it would require updating all the capsules again, and the current content of this PR has been pretty heavily reviewed.

File: kernel/src/grant.rs
Comment (lschuermann) [reply kernel/src/grant.rs:879]: Ah, sorry! Should've been "can this be a `slice::from_raw_parts`" (i.e. non-mutable) here? Not a safety concern per se.

File: kernel/src/grant.rs
Comment (bradjc) [reply kernel/src/grant.rs:218]: Does a capsule have any recourse if a schedule() fails? I'm guessing that is why this has escaped the resultifying this long, I'm not sure what a capsule could possibly do that would be worth it. Has any capsule ever even handled the bool?

I'm not saying we want to stick with the c convention, but we also probably don't want `upcalls.schedule().map_err(debug!("i have no idea what happened"))`.

**Comment by bradjc** [comment]
Addressed all of the comments to this point.

File: kernel/src/grant.rs
Comment (ppannuto) [reply kernel/src/grant.rs:218]: I don't have a compelling in-tree example I don't think, but my guess was that it was more likely that some capsule might need to do something extra if it thinks the process has died (e.g. check in some other way if the process is truly dead after the subscribe fails and choose to power off the associated peripheral).

It doesn't need the whole `map_err` bit, literally appending `.ok()` as shown is sufficient; that is a "Rust trick" I read somewhere along the way that just turns a `Result` into an `Option` (you lose the `Error` when you do this; Success -> `Some`, any error -> `None`), and the `Option` can be dropped without the compiler yelling.

@hudson-ayers follow-up PR seems fine; should be a pretty mechanical change.

**Review by ppannuto — APPROVED** [review]

File: kernel/src/grant.rs
Comment (bradjc) [reply kernel/src/grant.rs:218]: Thinking about this a little more, it's impossible for .schedule() to detect a dead process, because to enter a grant the process must be valid and while the grant is entered the process must stay valid. So schedule could fail because:

1. the `subscribe_num` was invalid (i.e. too big)
2. the upcall is null, i.e. the process never called subscribe
3. there is no room in the tasks array for the process

We don't need a persistent runtime check for 1, number 2 is normal operation that capsules aren't expected to defend against (I'm pretty sure), and for 3 its not clear what the capsule should do. Perhaps encapsulating these three cases in an Error type would spur someone to think about this any maybe come up with something.

And `ok()` is the clear tool for capsule authors to use, but I was just suggesting that if schedule() returns a proper result and a capsule author is thinking about what to do with the error case (i.e. "why does the kernel return me an error if I am not expected to do anything with it?"), they likely won't come up with anything and might write a debug statement instead.

**Review by hudson-ayers — APPROVED** [review]

**Review by lschuermann — APPROVED** [review]

**Comment by hudson-ayers** [comment]
I have tested the current head of this PR using several apps on Imix, including the `Imix`, `blink`, and `hello_loop` apps simultaneously.

bors r+

File: kernel/src/grant.rs
Comment (ppannuto) [reply kernel/src/grant.rs:218]: Opened #2656

Comment (phil-levis) [reply kernel/src/grant.rs:218]: If `schedule` cannot detect a dead process, what are the error cases?

It seems like this should return nothing. A `Result` is just going to add a lot of boilerplate that doesn't do anything.

**Pull request merged**
